gym-unity/gym_unity/envs/__init__.py

```diff
@@ -344,6 +358,79 @@ def _check_agents(self, n_agents: int) -> None:
                 "initialization. This is not supported."
             )
 
+    def _sanitize_info(self, info: BatchedStepResult) -> BatchedStepResult:
+        self._stored_info = info  # store the new original
+        if info.n_agents() == self._n_agents:
+            return info
+        n_extra_agents = info.n_agents() - self._n_agents
+        if self._n_agents is None:
+            self._n_agents = 1
+        if n_extra_agents < 0 or n_extra_agents > 2 * self._n_agents:
+            raise UnityGymException(
+                "The number of agents in the scene does not match the expected number."
+            )
+
+        # remove the done Agents
+        indices_to_keep: List[int] = []
+        for index in range(len(info.agent_id)):
+            if not info.done[index]:
+                indices_to_keep += [index]
+
+        # set the new AgentDone flags to True
+        for index in range(len(info.agent_id)):
+            agent_id = info.agent_id[index]
+            if agent_id not in self._stored_info.agent_id:
+                info.done[index] = True
+            if agent_id in self._done_agents:
+                info.done[index] = True
+        self._done_agents = []
+
+        _mask: Optional[List[np.array]] = None
+        if info.action_mask is not None:
+            _mask = []
+            for mask_index in range(len(info.action_mask)):
+                _mask += [info.action_mask[mask_index][indices_to_keep]]
+        new_obs: List[np.array] = []
+        for obs_index in range(len(info.obs)):
+            new_obs += [info.obs[obs_index][indices_to_keep]]
+        return BatchedStepResult(
+            obs=new_obs,
+            reward=info.reward[indices_to_keep],
+            done=info.done[indices_to_keep],
+            max_step=info.max_step[indices_to_keep],
+            agent_id=info.agent_id[indices_to_keep],
+            action_mask=_mask,
+        )
+
+    def _sanitize_action(self, action: np.array) -> np.array:
+        if self._stored_info.n_agents() == self._n_agents:
+            return action
+        sanitized_action = np.zeros(
+            (self._stored_info.n_agents(), self.group_spec.action_size)
+        )
+        input_index = 0
+        for index in range(self._stored_info.n_agents()):
+            if not self._stored_info.done[index]:
+                sanitized_action[index, :] = action[input_index, :]
+                input_index = input_index + 1
+        return sanitized_action
+
+    def _step(self, needs_reset: bool = False) -> BatchedStepResult:
+        if needs_reset:
+            self._env.reset()
+        else:
+            self._env.step()
+        info = self._env.get_step_result(self.brain_name)
+        while info.n_agents() < self._n_agents:
```

Comment:
Comment on the scary-looking while loop maybe

Reply:
Let me know what you think

Reply:
Should we add to the documentation that the gym can only be used with environments in which the agents request decisions at the same time and not to use the offset feature?